include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -1078,29 +1080,107 @@ void RunInParallelSection(ThreadPoolParallelSection &ps,
-  SummonWorkers(*pt, ps, n, fn);
-  profiler_.LogEndAndStart(ThreadPoolProfiler::DISTRIBUTION);
+  // index of thread that dispatch work to all other threads
+  int dispatch_q_idx = -1;
+  int dispatch_w_idx = -1;  // index of enqueued work
+  std::atomic<bool> dispatch_done{false};
+  std::atomic<bool> dispatch_work_done{false};
 
-  // Run work in the main thread
-  fn(0);
+  Task worker_fn = [&]() {
+    fn(ps.worker_idx.fetch_add(1, std::memory_order_relaxed));
+    ps.tasks_finished.fetch_add(1, std::memory_order_release);
+  };
+
+  std::function<int(int, Task&, bool)> enqueue_fn = [&, this](int q_idx, Task& w, bool logging) {
+    WorkerData& td = worker_data_[q_idx];
+    Queue& q = td.queue;
+    unsigned w_idx;
+    Task t = q.PushBackWithTag(w, pt->tag, w_idx);
+    if (t) {
+      return -1;
+    } else {
+      td.EnsureAwake();
+      if (logging) {
+        ps.tasks.push_back({q_idx, w_idx});
+      }
+      return static_cast<int>(w_idx);
+    }
+  };
+
+  unsigned current_dop = static_cast<unsigned>(ps.tasks.size()) + 1;
+  unsigned extra_needed = n > current_dop ? n - current_dop : 0;
+  std::vector<unsigned> good_hints, alt_hints;
+
+  Task dispatch_task = [&]() {  // dispatch work to other threads
+    int q_idx;
+    if (enable_good_worker_hints_) {
+      for (auto i = 1u; i < extra_needed; i++) {
+        if (i < good_hints.size()) {
+          q_idx = good_hints[i];
+        } else {
+          auto alt_i = i - static_cast<unsigned>(good_hints.size());
+          if (alt_i < alt_hints.size()) {
+            q_idx = alt_hints[alt_i];
+          } else {
+            q_idx = Rand(&pt->rand) % num_threads_;
+          }
+        }  // else
+        enqueue_fn(q_idx, worker_fn, true);
+      }
+    } else {
+      for (auto i = 1u; i < extra_needed; i++) {
+        enqueue_fn((dispatch_q_idx + i) % num_threads_, worker_fn, true);
+      }
+    }
+    dispatch_done.store(true, std::memory_order_release);  // dispatch complete
+    fn(ps.worker_idx.fetch_add(1, std::memory_order_relaxed));
+    dispatch_work_done.store(true, std::memory_order_release);  // work complete
+  };
+
+  if (extra_needed > 0) {
+    profiler_.LogStart();
+    if (enable_good_worker_hints_) {
+      GetGoodWorkerHints(extra_needed, good_hints, alt_hints);
+      if (!good_hints.empty()) {
+        dispatch_q_idx = good_hints[0];
+      } else if (!alt_hints.empty()) {
+        dispatch_q_idx = alt_hints[0];
+      } else {
+        dispatch_q_idx = Rand(&pt->rand) % num_threads_;
+      }
+    } else {
+      dispatch_q_idx = Rand(&pt->rand) % num_threads_;
+    }
+    dispatch_w_idx = enqueue_fn(dispatch_q_idx, dispatch_task, false);  // dispatch asynchronously
+    profiler_.LogEnd(ThreadPoolProfiler::DISTRIBUTION_ENQUEUE);
+  }
+
+  profiler_.LogEndAndStart(ThreadPoolProfiler::DISTRIBUTION);
+  fn(ps.worker_idx.fetch_add(1, std::memory_order_relaxed));  // run work in the main thread
```

Comment:
Suggest running fn(0) here, and allocating IDs to other worker threads from 1 upwards.  That provides some additional cache locality (when ID X from one loop is tending to access common data with ID X from previous loops), and makes a slight simplification on the logic.

Reply:
Done.